### PR TITLE
Parse assertions with Holder-of-Key profile

### PR DESCRIPTION
**Pull request by abliqo**
# Introduction

Holder-of-Key assertions are used to achieve higher levels of federation security, compared to bearer assertions, by having Relying Party challenge subscriber to prove possession of the key specified in the assertion that represents subscriber in addition to verifying the assertion itself signed by Identity Provider. More information about it can be found in https://pages.nist.gov/800-63-3/sp800-63c.html

# Problem

This library fails to parse SAML responses containing assertions with Holder-of-Key profile, for example:
```
<ns1:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:holder-of-key">
    <ns1:SubjectConfirmationData InResponseTo="id-KHlas49TtW2VdC8WN" NotOnOrAfter="2019-05-14T20:36:13Z" Recipient="https://sp:443/.auth/saml/login">
        <ns2:KeyInfo>
            <ns2:X509Data>
                <ns2:X509Certificate>MIICITCCAYoCAQEwDQYJKoZIhvcNAQELBQAwWDELMAkGA1UEBhMCenoxCzAJBgNVBAgMAnp6MQ0wCwYDVQQHDAR6enp6MQ4wDAYDVQQKDAVaenp6ejEOMAwGA1UECwwFWnp6enoxDTALBgNVBAMMBHRlc3QwIBcNMTkwNDEyMTk1MDM0WhgPMzAxODA4MTMxOTUwMzRaMFgxCzAJBgNVBAYTAnp6MQswCQYDVQQIDAJ6ejENMAsGA1UEBwwEenp6ejEOMAwGA1UECgwFWnp6enoxDjAMBgNVBAsMBVp6enp6MQ0wCwYDVQQDDAR0ZXN0MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDHcj80WU/XBsd9FlyQmfjPUdfmedhCFDd6TEQmZNNqP/UG+VkGa+BXjRIHMfic/WxPTbGhCjv68ci0UDNomUXagFexLGNpkwa7+CRVtoc/1xgq+ySE6M4nhcCutScoxNvWNn5eSQ66i3U0sTv91MgsXxqEdTaiZg0BIufEc3dueQIDAQABMA0GCSqGSIb3DQEBCwUAA4GBAGUV5B+USHvaRa8kgCNJSuNpo6ARlv0ekrk8bbdNRBiEUdCMyoGJFfuM9K0zybX6Vr25wai3nvaog294Vx/jWjX2g5SDbjItH6VGy6C9GCGf1A07VxFRCfJn5tA9HuJjPKiE+g/BmrV5N4CealzFxPHWYkNOzoRU8qI7OqUai1kL</ns2:X509Certificate>
            </ns2:X509Data>
        </ns2:KeyInfo>
    </ns1:SubjectConfirmationData>
</ns1:SubjectConfirmation>
```

fails to be parsed with the following error:
```
ERROR    saml2.response:response.py:836 get subject
Traceback (most recent call last):
  File "/home/abublich/repos/abliqo-pysaml2/venv/local/lib/python2.7/site-packages/pysaml2-4.7.0-py2.7.egg/saml2/response.py", line 828, in _assertion
    self.get_subject()
  File "/home/abublich/repos/abliqo-pysaml2/venv/local/lib/python2.7/site-packages/pysaml2-4.7.0-py2.7.egg/saml2/response.py", line 753, in get_subject
    if not self._holder_of_key_confirmed(_data):
  File "/home/abublich/repos/abliqo-pysaml2/venv/local/lib/python2.7/site-packages/pysaml2-4.7.0-py2.7.egg/saml2/response.py", line 730, in _holder_of_key_confirmed
    [samlp, saml, xenc, ds]):
  File "/home/abublich/repos/abliqo-pysaml2/venv/local/lib/python2.7/site-packages/pysaml2-4.7.0-py2.7.egg/saml2/__init__.py", line 1004, in extension_elements_to_elements
    for extension_element in extension_elements:
TypeError: 'SubjectConfirmationData' object is not iterable
```

The root cause is two-fold:
1. The type SubjectConfirmationDataType_ does not declare KeyInfo as child element.
2. The bug in function _holder_of_key_confirmed: it should check KeyInfo child element of SubjectConfirmationData instead of SubjectConfirmationData itself.

# Solution

Fixed the root cause and added new unit tests that verify successful parsing of Holder-of-Key assertions.

### CODE REVIEW

**Comment by abliqo** [comment]
@c00kiemon5ter , @Gijutsu , @jkakavas  could you please kindly review & merge this change?

File: src/saml2/saml.py [hunk 1]
```diff
@@ -496,6 +500,7 @@ def __init__(self,
                           text=text,
                           extension_elements=extension_elements,
                           extension_attributes=extension_attributes)
+        self.key_info = key_info
```
Comment (c00kiemon5ter) [inline src/saml2/saml.py]: `KeyInfo` is not part of the `SubjectConfirmationDataType` spec. It is only one of the `xs:any` elements that are allowed to extend the `SubjectConfirmationData` element. Extensions to elements are part of `extension_elements`.

File: src/saml2/response.py [hunk 2]
```diff
@@ -722,11 +722,12 @@ def _bearer_confirmed(self, data):
...
     def _holder_of_key_confirmed(self, data):
-        if not data:
+        if not data or not data.key_info:
             return False
 
         has_keyinfo = False
-        for element in extension_elements_to_elements(data,
+        key_info = data.key_info or ()
+        for element in extension_elements_to_elements(key_info,
                                                       [samlp, saml, xenc, ds]):
```
Comment (c00kiemon5ter) [inline src/saml2/response.py]: The needed change here is

```py
for element in extension_elements_to_elements(
    data.extension_elements, [samlp, saml, xenc, ds]
):
    ...
```

`extension_elements_to_elements()` will map the extension elements to objects (ie, in this case an array with an element of `ds.KeyInfo` will be returned)

Comment (abliqo) [reply src/saml2/response.py]: Thanks. I changed the code to use `extension_elements` instead.

File: src/saml2/saml.py (hunk 1 above)
Comment (abliqo) [reply src/saml2/saml.py]: Thanks. Done.

**Comment by c00kiemon5ter** [comment]
Thank you

**Pull request merged**